src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.BoolValueSet.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    using static BinaryOperatorKind;
+
+    internal static partial class ValueSetFactory
+    {
+        private class BoolValueSet : IValueSet<bool>
+        {
+            private readonly bool _hasFalse, _hasTrue;
+
+            internal static BoolValueSet AllValues = new BoolValueSet(hasFalse: true, hasTrue: true);
+            internal static BoolValueSet None = new BoolValueSet(hasFalse: false, hasTrue: false);
+            internal static BoolValueSet OnlyTrue = new BoolValueSet(hasFalse: false, hasTrue: true);
+            internal static BoolValueSet OnlyFalse = new BoolValueSet(hasFalse: true, hasTrue: false);
+
+            private BoolValueSet(bool hasFalse, bool hasTrue) => (_hasFalse, _hasTrue) = (hasFalse, hasTrue);
+
+            public static BoolValueSet Create(bool hasFalse, bool hasTrue)
+            {
+                switch (hasFalse, hasTrue)
+                {
+                    case (false, false):
+                        return None;
+                    case (false, true):
+                        return OnlyTrue;
+                    case (true, false):
+                        return OnlyFalse;
+                    case (true, true):
+                        return AllValues;
+                }
+            }
+
+            bool IValueSet.IsEmpty => !_hasFalse && !_hasTrue;
+
+            IValueSetFactory<bool> IValueSet<bool>.Factory => BoolValueSetFactory.Instance;
+
+            IValueSetFactory IValueSet.Factory => BoolValueSetFactory.Instance;
+
+            public bool Any(BinaryOperatorKind relation, bool value)
+            {
+                switch (relation, value)
+                {
+                    case (Equal, true):
```

Comment:
>Equal [](start = 26, length = 5)

What about the `NotEqual` operator? 
Speclet says "Additionally, we support == and != for string and bool."

Do we have `CompileAndVerify` tests on these permutations?

Reply:
I initially supported them but them removed support later.  I'll update the spec to remove them.  We already have `==` support through the constant pattern and `!=` support through the `not` pattern.  Supporting `!=` would result in some pretty unintuitive behavior, for example `!= 3` would mean that the input is an `int` that is equal to some value other than `3`.  An input of type string does not match `!= 3`!  For this and other reasons (don't get me started on `double.NaN`) I think it is best not to support NotEquals.  If you really intended what `!= 3` would mean, you would write `int and not 3`, which I think is much more intuitive.  If the input is already of type `int` then you need only write `not 3`.


---
In reply to: [381018394](https://github.com/dotnet/roslyn/pull/41503#discussion_r381018394) [](ancestors = 381018394)